lua/lualine/components/branch.lua

```diff
@@ -29,22 +29,22 @@ timer:start(0, 1000, vim.schedule_wrap(function()
   end
 end))
 
-local function branch()
-  if not git_branch or #git_branch == 0 then return '' end
-  local ok,devicons = pcall(require,'nvim-web-devicons')
-  if ok then
-    local icon = devicons.get_icon('git')
-    if icon ~= nil then
-      return icon .. ' ' .. git_branch
+local function branch(args)
+  local icon = '' -- e0a0
+  if args.icons_enabled == false then
+    icon = nil
+  elseif args.icons_enabled == nil and args.opt.icons_enabled == false then
```

Comment:
This is bad imo, why have both `args` and `args.opt`? 

There should only be one parameter here `options` or `opts`. The options should be parsed when loading the component. Something like
```
component.options = M.options
for name, value in (component options array)
  component.options.name = value
```

This will make components more clear imo. I will look into this pr more when this is updated.

Reply:
Theres args.opt.icons_enabled and args.icons_enabled . Brecause args.opt is global options. and args is component specific options .

Users can set ```lualine.opt.icons_enabled = false```
like users set ```lualine.seperator = '|'```
to not have icons in any components insted of having to set it on every component . And then can override that global option in components. 
Like if users wants icons in filetype and nowhere else they can do something like
```lualine.section_c = { {'filetype', icons_enabled = true}}```

Also not all options are processed in compinents . global options like padding is handled when components are drawn in util.lua

Reply:
> Theres args.opt.icons_enabled and args.icons_enabled . Brecause args.opt is global options. and args is component specific options .

Well, what exactly is the benefit of having them both separately? I cannot see one. They should be parsed when loading the component for sure.

> Users can set `lualine.opt.icons_enabled = false`
> like users set `lualine.seperator = '|'`
> to not have icons in any components insted of having to set it on every component . And then can override that global option in components.
> Also not all options are processed in compinents . global options like padding is handled when components are drawn in util.lua

This is inconsistent as hell. First of all why are some options in `opt` and others are not? Secondly I don't see a reason why padding cannot be a component option, it surely can.

Imo all options should be in the `opt` object ( I would personally rename it to options though).

This way the user can set everything through `lualine.options`

The component loading function should be aware (probably hard coded) which of these global options to pass to components (icons should be passed for example) and update them if the given component has an overriding option

Reply:
I'm probably not making my self clear.
Here's my lualine conf.

```lua
vim.cmd("packadd lualine.nvim")
vim.cmd("packadd nvim-web-devicons")
local lualine = require('lualine')

vim.o.showmode = false
lualine.separator = '|'
lualine.opt.icons_enabled = true

lualine.sections = {
  lualine_a = { {'mode', lower=false} },
  lualine_b = { {'branch', icons_enabled=true}, {'signify', colored=true}}, 
  lualine_c = { {'filename', modified=true, padding = 1, full_name=true, relative=true}},
  lualine_x = { 'encoding', {'fileformat', icons_enabled=true}, {'filetype', icons_enabled=true }},
  lualine_y = { 'progress' },
  lualine_z = { {'location', icons_enabled=false}  },
  lualine_diagnostics = {  }
}

lualine.status()
```

opt table is not something user will set in components. 
Also most of the options i've set here are in default values so they aren't required to be mentioned. I've put them for testing purposes.

>Well, what exactly is the benefit of having them both separately? 

They allow user to set an option globally . The options that effect more than one component like `icons_enabled`

> Secondly I don't see a reason why padding cannot be a component option, it surely can.

Options like padding, lower, upper are same for all components no point in implementing the samething on every component.

>This way the user can set everything through lualine.options

That will be like current opt table . In that case users wonn't be able to set different value for an option for different components.

>The component loading function should be aware (probably hard coded) which of these global options to pass to components (icons should be passed for example) and update them if the given component has an overriding option

please elaborate . I couldn't understand this part.

If you haven't already please look at the entire thing when you have time .

Reply:
I understand that everything works well now, I have tested it on my system. I'm talking about an implementation... I tried my best to explain it, I will try to clean this up myself soon and you will be able to see what I meant :)

Reply:
Alright .